compiler/rustc_ast_passes/src/feature_gate.rs

```diff
@@ -342,6 +344,12 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
             ast::TyKind::Never => {
                 gate!(&self, never_type, ty.span, "the `!` type is experimental");
             }
+            ast::TyKind::Path(_, x) if x == &Symbol::intern("f16") => {
+                gate!(&self, f16, ty.span, "the f16 primitive type is experimental");
+            }
+            ast::TyKind::Path(_, x) if x == &Symbol::intern("f128") => {
+                gate!(&self, f128, ty.span, "the f128 primitive type is experimental");
+            }
```

Comment:
I added a gate here but this doesn't seem right since it can't catch inferred types. I'm not sure where to do a check after types are resolved.

Reply:
Well, you need to specify the type somewhere to be able to infer it, so the gate will apply at the definition place. I think constructing a `f128` via inference if the type is specified in a different crate is the only possible problem here, and I'm not sure how much of an issue that is since the `feature(...)` still has to exist somewhere in the crate graph.

Reply:
@compiler-errors do you have any suggestion here?

Reply:
the gating is also incorrect the other way, it will deny valid code like `struct f16; f16;`

Reply:
Ah, that is unfortunate. What is the better way to do this?